File: tools/unown_forms.py
"""Unown letter forms for the sprite build tools.

Unown is one species (dex 201) with 26 letters. The HGSS/PokeMMO follow-sprite art has one sheet per letter:
    201-b-n.png        letter A      201-b-s.png        (shiny)
    201-b-n-01.png     letter B      201-b-s-01.png
    ...
    201-b-n-25.png     letter Z      201-b-s-25.png
(-26 and -27 are "!" and "?", which Gold does not have.)

The runtime pipeline is keyed by a numeric species id, so each letter B..Z is baked as an extra "form asset" with id
FORM_BASE + n (60001..60025) through the same generators as a species. lib/unown_forms.lua is the runtime side and must
agree on FORM_BASE / FORM_COUNT. The entries are synthesized here rather than written into the followsprites mapping
(tools/generate_followsprites_mapping.ps1 owns that file and would drop them on regeneration).
"""
from __future__ import annotations

import copy
from pathlib import Path

BASE_DEX = 201
FORM_BASE = 60000
FORM_COUNT = 25
SRC_REL = "assets/enhanced_overworld/followsprites"


def is_form_id(dex: int) -> bool:
    return FORM_BASE < int(dex) <= FORM_BASE + FORM_COUNT


def source_files(dex: int) -> dict[str, str]:
    """{'normal': '201-b-n-03.png', 'shiny': '201-b-s-03.png'} for a form id, {} otherwise."""
    if not is_form_id(dex):
        return {}
    n = int(dex) - FORM_BASE
    return {"normal": f"{BASE_DEX:03d}-b-n-{n:02d}.png", "shiny": f"{BASE_DEX:03d}-b-s-{n:02d}.png"}


def mapping_entries(species: dict, root: Path) -> dict[str, dict]:
    """Synthesized followsprites-mapping entries (same shape as species['201']) for every letter whose art exists."""
    base = species.get(str(BASE_DEX))
    if not isinstance(base, dict):
        return {}
    out: dict[str, dict] = {}
    for n in range(1, FORM_COUNT + 1):
        sid = FORM_BASE + n
        files = source_files(sid)
        entry = copy.deepcopy(base)
        entry["speciesId"] = sid
        ok = True
        for variant, fname in files.items():
            block = entry.get(variant)
            if not isinstance(block, dict) or not (root / SRC_REL / fname).is_file():
                ok = False
                break
            block["file"] = fname
            block["path"] = f"{SRC_REL}/{fname}"
        if ok:
            out[str(sid)] = entry
    return out

File: lib/unown_forms.lua
-- Unown letter forms (Gold). Unown is one species with 26 letters; the engine reads the letter off the DVs
-- (src/core/gen2/Unown.lua: letterFromDVs, 1 = A .. 26 = Z). Our HGSS/PokeMMO art has one sheet per letter
-- (assets/enhanced_overworld/followsprites/201-b-n[-NN].png: no suffix = A, -01 = B .. -25 = Z), which the build bakes
-- as extra "form assets" with ids 60001..60025 (tools/unown_forms.py) so every id-keyed part of the sprite pipeline
-- (runtime sheets, True Size packs, the atlas) serves them unchanged. 60000+ never collides with a real species
-- (1..1025) or a dex-expansion form (30001..50248).
--
-- A form is the number in the file suffix (1..25); nil means the base art (letter A, or any non-Unown).
local UnownForms = {}

UnownForms.SPECIES = "UNOWN"
UnownForms.BASE_DEX = 201
UnownForms.FORM_BASE = 60000
UnownForms.FORM_COUNT = 25 -- B..Z; Gold has no "!" / "?"

--- Letter number (1 = A .. 26 = Z) -> form 1..25, or nil for A / anything out of range.
function UnownForms.formForLetter(letter)
  local n = tonumber(letter)
  if not n or math.floor(n) ~= n then return nil end
  if n < 2 or n > UnownForms.FORM_COUNT + 1 then return nil end
  return n - 1
end

--- Form 1..25 -> letter number 2..26 (nil when out of range).
function UnownForms.letterForForm(form)
  local n = tonumber(form)
  if not n or math.floor(n) ~= n or n < 1 or n > UnownForms.FORM_COUNT then return nil end
  return n + 1
end

--- Form asset id (60001..60025) for a form, or nil.
function UnownForms.assetId(form)
  local n = tonumber(form)
  if not n or math.floor(n) ~= n or n < 1 or n > UnownForms.FORM_COUNT then return nil end
  return UnownForms.FORM_BASE + n
end

function UnownForms.isFormId(id)
  local n = tonumber(id)
  return n ~= nil and math.floor(n) == n
    and n > UnownForms.FORM_BASE and n <= UnownForms.FORM_BASE + UnownForms.FORM_COUNT
end

--- 60003 -> 201 (the species whose art/scale the form shares); any other id comes back unchanged.
function UnownForms.baseId(id)
  if UnownForms.isFormId(id) then return UnownForms.BASE_DEX end
  return id
end

--- Form asset id for `dex` (an asset id) wearing `form`, or nil when this is not Unown or the form is not a letter B..Z.
function UnownForms.formAssetId(dex, form)
  if tonumber(dex) ~= UnownForms.BASE_DEX then return nil end
  return UnownForms.assetId(form)
end

return UnownForms
